gnuradio-runtime/include/gnuradio/block.h

```diff
@@ -180,6 +180,31 @@ namespace gr {
      */
     virtual bool stop();
 
+    /*!
+     * \brief Called to perform any post-constructor initialization.
+     *
+     * This allows a block to perform any setup that may need to happen
+     * which cannot be performed in the constructor. This will be called
+     * during flat_flowgraph::setup_connections()
+     */
+    virtual bool init();
+
+    /*!
+     * \brief Called to perform buffer allocation for a specific input port.
+     *
+     * This allows a block to allocate the buffer, but only if the block
+     * indicated it wants to allocate buffers in it's io_signature.
+     */
+    virtual buffer_sptr allocate_input_buffer(int port);
```

Comment:
Consider renaming this to 'allocate_upstream_output_buffer', as there aren't actually any input buffers.